src/main/java/com/google/api/gax/grpc/ProviderManager.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2016, Google Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.grpc;
+
+import io.grpc.ManagedChannel;
+import java.io.IOException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * ProviderManager gives a way to manage the lazy creation of channels and executors that
+ * are shared among multiple API wrappers.
+ *
+ * <pre>
+ * <code>
+ *   ProviderManager providerManager =
+ *     ProviderManager.newBuilder()
+ *       .setChannelProvider(SomeApiSettings.defaultChannelProviderBuilder().build())
+ *       .setExecutorProvider(SomeApiSettings.defaultExecutorProviderBuilder().build())
+ *       .build();
+ *
+ *   SomeApiSettings settingsA =
+ *     SomeApiSettings.defaultBuilder()
+ *       .setExecutorProvider(providerManager)
+ *       .setChannelProvider(providerManager)
+ *       .build();
+ *   SomeApiSettings settingsB =
+ *     SomeApiSettings.defaultBuilder()
+ *       .setExecutorProvider(providerManager)
+ *       .setChannelProvider(providerManager)
+ *       .build();
+ *
+ *   providerManager.shutdown();
+ * </code>
+ * </pre>
+ */
+public class ProviderManager implements ChannelProvider, ExecutorProvider {
```

Comment:
Ahh, yes, that is what will happen. And actually, both channels will be closed in this scenario; the executor will not be closed (which is correct for FixedExecutorProvider).